sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -700,6 +684,34 @@ private[sql] object RelationalGroupedDataset {
     new RelationalGroupedDataset(df, groupingExprs, groupType: GroupType)
   }
 
+  private[sql] def handleGroupingExpression(
+      logicalPlan: LogicalPlan,
+      sparkSession: SparkSession,
+      groupingExprs: Seq[Expression]): (QueryExecution, Seq[Attribute]) = {
+    // Resolves grouping expressions.
+    val dummyPlan = Project(groupingExprs.map(alias), LocalRelation(logicalPlan.output))
+    val analyzedPlan = sparkSession.sessionState.analyzer.execute(dummyPlan)
+      .asInstanceOf[Project]
+    sparkSession.sessionState.analyzer.checkAnalysis(analyzedPlan)
+    val aliasedGroupings = analyzedPlan.projectList
+
+    // Adds the grouping expressions that are not in base DataFrame into outputs.
+    val addedCols = aliasedGroupings.filter(g => !logicalPlan.outputSet.contains(g.toAttribute))
+    val qe = Dataset.ofRows(
+      sparkSession,
+      Project(logicalPlan.output ++ addedCols, logicalPlan)).queryExecution
+
+    (qe, aliasedGroupings.map(_.toAttribute))
+  }
+
+  private def alias(expr: Expression): NamedExpression = expr match {
```

Comment:
I mean this is sort of like the named functionality, but different.

Reply:
This adds `case u: UnresolvedFunction => UnresolvedAlias(expr, None)` mostly. Not sure why this is a special case here.

Regarding  `TypedAggregate` logic, we can move it out around each groupingExprs (a few places, a bit out of the scope of this PR) and aggExprs (easy).

Reply:
This is soooo subtle, I moved the aggExpr a bit earlier, then everything breaks.

Reply:
We should follow up on this.